tests/Operation/FindOneAndReplaceFunctionalTest.php

```diff
@@ -0,0 +1,66 @@
+<?php
+
+namespace MongoDB\Tests\Operation;
+
+use MongoDB\Driver\BulkWrite;
+use MongoDB\Operation\FindOneAndReplace;
+use MongoDB\Tests\CommandObserver;
+use MongoDB\Tests\Fixtures\Codec\TestDocumentCodec;
+use MongoDB\Tests\Fixtures\Document\TestObject;
+
+class FindOneAndReplaceFunctionalTest extends FunctionalTestCase
+{
+    public function testFindAndReplaceOneWithCodec(): void
+    {
+        $this->createFixtures(1);
+
+        (new CommandObserver())->observe(
+            function (): void {
+                $replaceObject = TestObject::createForFixture(1);
+                $replaceObject->x->foo = 'baz';
+
+                $operation = new FindOneAndReplace(
+                    $this->getDatabaseName(),
+                    $this->getCollectionName(),
+                    ['_id' => 1],
+                    $replaceObject,
+                    ['codec' => new TestDocumentCodec()],
+                );
+
+                $this->assertEquals(
+                    TestObject::createForFixture(1, true),
+                    $operation->execute($this->getPrimaryServer()),
+                );
+            },
+            function (array $event): void {
+                $this->assertEquals(
+                    (object) [
+                        '_id' => 1,
+                        'x' => (object) ['foo' => 'baz'],
+                        'encoded' => true,
+                    ],
+                    $event['started']->getCommand()->update ?? null,
+                );
+            },
+        );
+    }
+
+    /**
+     * Create data fixtures.
+     */
+    private function createFixtures(int $n): void
+    {
+        $bulkWrite = new BulkWrite(['ordered' => true]);
+
+        for ($i = 1; $i <= $n; $i++) {
+            $bulkWrite->insert([
+                '_id' => $i,
+                'x' => (object) ['foo' => 'bar'],
+            ]);
```

Comment:
This seems closely coupled with TestObject. Did you consider inserting `TestObject::createForFixture($i)` here instead? That may require a Serializable implementation to ensure `$id` gets mapped to an `_id` field.

I'm not sure if there are other places where you've coupled the structure of TestObject in a fixture method, but the same suggestion would apply there as well if so.

Reply:
Updated to create a document through a factory instead. I've also updated other instances of that same fixture method in these tests.